### ISSUE TITLE
posix_fallocate is not signal-safe

### ISSUE THREAD

**Issue by tomaszkapela** [pmem/issues#168]
A call to `posix_fallocate` can be interrupted by a signal and then the whole operation is cancelled and errno is set to EINTR. If an application using NVML has a lot of signals flying around, the file allocation might be impossible. This issue should be addressed in one way or another.

**Comment by krzycz** [comment]
Seems like similar problem may occur in case of other system calls (i.e. `read`).

**Comment by marcinslusarz** [comment]
Yeah, I looked briefly at the syscalls we use and these can be interrupted: close, flock, dup, read.

**Comment by vitalyvch** [comment]
By the fact there, in NVML, are much many syscalls which are sensitive to signals.

**Comment by GBuella** [comment]
I looked at this, and I see os_posix_fallocate used in only three places. In all three instances, the error code seems to be propagated to the user in `errno`. Some pmem* functions can fail with EINTR Should we add this to the documentation?
If the application using NVML uses signals for some communication, it needs to handle its own `read` ,`dup` etc. calls anyways, so it let it just handle these interruptions the same way.

**Comment by marcinslusarz** [comment]
The problem is that restarting pmemobj_create does not help. Every time we do it posix_fallocate starts from the very beginning, which means that it might never finish if application raises signals periodically (like SIGALRM).

**Comment by GBuella** [comment]
Yes, the same is true for whatever syscall that application does. If the application itself calls fallocate on some file, it will need to deal with EINTR.
And the application perhaps doesn't want NVML to continue allocating at all, perhaps the alarm signal is used to communicate to the application, that something other than opening the pool should be done. It might get alarm signals over and over again to signal "don't bother with pmem pool, do this other urgent thing instead".
On the other hand, if the application wants to make sure syscalls don't get interrupted a lot in some code section, it can just stop communicating with signals for that time, or just temporarily block those signals using thread specific signal mask.

**Comment by GBuella** [comment]
In short: you write a library, you didn't raise a signal, you *don't know why there was a signal*.

**Comment by vitalyvch** [comment]
@marcinslusarz , You can play with block/unblock to prevent it.

**Comment by marcinslusarz** [comment]
We should not block signals, because that may interfere with the way application handles signals.
I think we should have a loop around posix_fallocate and iterate over small enough blocks for one syscall to complete.
Something like:
```
for (off = 0; off < size; off += blksize)
  while (posix_fallocate(fd, off, blksize) == EINTR)
      ;

```

**Comment by vitalyvch** [comment]
> We should not block signals, because that may interfere with the way application handles signals.

Marcin, your words really disappoint me :-(

**Comment by marcinslusarz** [comment]
Fixed in PMDK 1.6
